src/ImageSharp/Diagnostics/MemoryDiagnostics.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Threading;
+
+namespace SixLabors.ImageSharp.Diagnostics
+{
+    /// <summary>
+    /// Represents the method to handle <see cref="MemoryDiagnostics.UndisposedAllocation"/>.
+    /// </summary>
+    public delegate void UndisposedMemoryResourceDelegate(string allocationStackTrace);
+
+    /// <summary>
+    /// Utilities to track memory usage and detect memory leaks from not disposing ImageSharp objects.
+    /// </summary>
+    public static class MemoryDiagnostics
+    {
+        private static int totalUndisposedAllocationCount;
+
+        private static UndisposedMemoryResourceDelegate undisposedMemoryResource;
+        private static int undisposedMemoryResourceSubscriptionCounter;
+
+        /// <summary>
+        /// Fires when an ImageSharp object's undisposed memory resource leaks to the finalizer.
+        /// The event brings significant overhead, and is intended to be used for troubleshooting only.
+        /// For production diagnostics, use <see cref="TotalUndisposedAllocationCount"/>.
+        /// </summary>
+        public static event UndisposedMemoryResourceDelegate UndisposedAllocation
+        {
+            add
+            {
+                Interlocked.Increment(ref undisposedMemoryResourceSubscriptionCounter);
+                undisposedMemoryResource += value;
```

Comment:
Instead of `Interlocked.Increment/Decrement` I'd put both operations under one `lock` that's also used in `remove` (the lock object).

`add`/`remove` isn't high perf, and so both operations are protected. Current implementation could suffer from a race.